include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,462 @@
+    template <typename VisitPolicy>
+    static bool has_correct_intersection_orientation(polygon_2d const& polygon,
+                                                     point_2d const& point1,
+                                                     point_2d const& point2,
+                                                     VisitPolicy& visitor)
+    {
+        auto ring = boost::geometry::exterior_ring(polygon);
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
+        }
+
+        for (auto it = boost::begin(ring); it != boost::end(ring); ++it)
+        {
+            if (boost::geometry::equals(*it, point1))
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
+                    return false;
+                }
+                auto prev = (it == boost::begin(ring)) ? boost::end(ring) - 1 : boost::prior(it);
+                if (boost::geometry::equals(point2, *prev))
+                {
+                    // valid intersection and inconsistent orientation
+                    return true;
+                }
+                auto next = (it == boost::end(ring)) ? boost::begin(ring) : boost::next(it);
+                if (boost::geometry::equals(point2, *next))
+                {
+                    // valid intersection but inconsistent orientation
+                    if (AllowInconsistentOrientation)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        visitor.template apply<failure_inconsistent_orientation>();
+                        return false;
+                    }
+                }
+            }
+        }
+        visitor.template apply<failure_invalid_intersection>();
+        return false;
+    }
+
+    // Check if a segment intersects a polygon in 3D space
+    template <typename VisitPolicy, typename Strategy>
+    intersection_type check_polygon_segment_intersection(point_3d const& point1,
+                                                         point_3d const& point2,
+                                                         polygon_3d const& polygon,
+                                                         VisitPolicy& visitor,
+                                                         Strategy const& strategy)
+    {
+        vec3 p1;
+        vec3 p2;
+        set_vec_from_point::apply(point1, p1);
+        set_vec_from_point::apply(point2, p2);
+
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+        polygon_2d projected_polygon;
+
+        // Check if the polygon has been projected before
+        std::ostringstream oss;
+        oss << boost::geometry::wkt(polygon);
+        if (m_polygon_projections.find(oss.str()) != m_polygon_projections.end())
```

Comment:
This seems rather expensive aside from pulling the wkt dependency into this. Can we use the indices of the polygons in their original datastracture as integer keys instead? (I understand that all boost geometry concepts require random access ranges).